ui-lib/src/main/java/co/electriccoin/zcash/ui/common/FlowExt.kt

```diff
@@ -0,0 +1,57 @@
+package co.electriccoin.zcash.ui.common
+
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import kotlin.time.Duration
+import kotlin.time.ExperimentalTime
+import kotlin.time.TimeSource
+
+@OptIn(ExperimentalTime::class)
+fun <T> Flow<T>.throttle(
+    duration: Duration,
+    timeSource: TimeSource = TimeSource.Monotonic
+): Flow<T> = flow {
+    coroutineScope {
+        val context = coroutineContext
+        val mutex = Mutex()
+
+        var timeMark = timeSource.markNow()
+        var delayEmit: Deferred<Unit>? = null
+        var firstValue = true
+        collect { value ->
+            if (firstValue) {
+                firstValue = false
+                emit(value)
+                timeMark = timeSource.markNow()
+                return@collect
+            }
+
+            if (timeMark.elapsedNow() >= duration) {
+                delayEmit?.cancel()
+                mutex.withLock {
+                    emit(value)
+                    timeMark = timeSource.markNow()
+                }
+            } else {
+                delayEmit?.cancel()
+                delayEmit = async(Dispatchers.Default) {
+                    mutex.withLock {
+                        delay(duration)
+                        withContext(context) {
+                            emit(value)
```

Comment:
I think we still have a bug to address: this emits potentially stale values.  The value is captured at the time of `collect` above but is then delayed.  We actually want to emit the most recently received value.  I think this implies having a `var` outside of the `collect` holding a reference to the most recently received value.

So the async isn't about delayed emission of a specific value but actually delayed emission of the latest value.